segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,483 @@
+
+            txn.update(oldChunk);
+        } else {
+            segmentMetadata.setFirstChunk(newChunkName);
+        }
+        segmentMetadata.setLastChunk(newChunkName);
+        segmentMetadata.setLastChunkStartOffset(offset);
+        segmentMetadata.checkInvariants();
+        // Save the segment metadata.
+        txn.update(segmentMetadata);
+    }
+
+    private String getSystemJournalFileName() {
+        return getSystemJournalFileName(containerId, epoch, currentFileIndex);
+    }
+
+    private String getSystemJournalFileName(int containerId, long epoch, long currentFileIndex) {
+        return NameUtils.getSystemJournalFileName(containerId, epoch, currentFileIndex);
+    }
+
+    private ChunkHandle getChunkHandleForSystemJournal() throws ChunkStorageException {
+        ChunkHandle h;
+        val systemLogName = getSystemJournalFileName();
+        if (!chunkStorage.exists(systemLogName)) {
+            h = chunkStorage.create(systemLogName);
+        } else {
+            h = chunkStorage.openWrite(systemLogName);
+        }
+        return h;
+    }
+
+    /**
+     * Apply truncate action to the segment metadata.
+     */
+    private void applyTruncate(MetadataTransaction txn, String segmentName, long truncateAt, long firstChunkStartsAt) throws StorageMetadataException {
+        SegmentMetadata segmentMetadata = (SegmentMetadata) txn.get(segmentName);
+        segmentMetadata.checkInvariants();
+        String currentChunkName = segmentMetadata.getFirstChunk();
+        ChunkMetadata currentMetadata;
+        long startOffset = segmentMetadata.getFirstChunkStartOffset();
+        while (null != currentChunkName) {
+            currentMetadata = (ChunkMetadata) txn.get(currentChunkName);
+            // If for given chunk start <= truncateAt < end  then we have found the chunk that will be the first chunk.
+            if ((startOffset <= truncateAt) && (startOffset + currentMetadata.getLength() > truncateAt)) {
+                break;
+            }
+
+            startOffset += currentMetadata.getLength();
+            // move to next chunk
+            currentChunkName = currentMetadata.getNextChunk();
+            txn.delete(currentMetadata.getName());
+        }
+        Preconditions.checkState(firstChunkStartsAt == startOffset);
+        segmentMetadata.setFirstChunk(currentChunkName);
+        segmentMetadata.setStartOffset(truncateAt);
+        segmentMetadata.setFirstChunkStartOffset(firstChunkStartsAt);
+        segmentMetadata.checkInvariants();
+
+    }
+
+    /**
+     * Indicates whether given segment is a system segment.
+     *
+     * @param segmentName Name of the sgement to check.
+     * @return True if given segment is a system segment.
+     */
+    public boolean isStorageSystemSegment(String segmentName) {
+        if (segmentName.startsWith(systemSegmentsPrefix)) {
+            for (String systemSegment: systemSegments) {
+                if (segmentName.equals(systemSegment)) {
```

Comment:
This method seems slow.

Reply:
In reality it checks if segment starts with `_system` prefix and then checks if it is one of the four segments.
But I think adding a flag in segment metadata itself makes more sense.

Reply:
Fixed. This code is not called now. Left behind just because method useful only for validation purposes.